nsot/models.py

```diff
@@ -1881,28 +1925,34 @@ def clean_interface(self, value):
         return value
 
     def clean_type(self, value):
+        """Ensure that ProtocolType matches our site."""
+        if self.site != value.site:
+            raise exc.ValidationError({
+                'type': 'The type must be on the same site as this Protocol'
+            })
+
+        return value
+
+    def set_attributes(self, attributes, valid_attributes=None, partial=False):
         """
         Ensure that all attributes are set that are required by the set
-        ProtocolType
+        ProtocolType.
         """
-        required = value.required_attributes.values_list(
-            'name', flat=True
-        )
-
-        # If we're receiving a new set of attributes, validate against those,
-        # otherwise use the ones we already have
-        if self._set_attributes:
-            current = set(self._set_attributes)
-        else:
-            current = set(self.get_attributes())
-        missing = set(required).difference(current)
+        required = self.type.get_required_attributes()
+        if valid_attributes is None:
+            valid_attributes = Attribute.all_by_name(
+                'Protocol', site=self.site
+            )
 
-        if missing:
-            raise exc.ValidationError({
-                'attributes': 'Missing required attributes: %r' % list(missing)
-            })
+        # Temporarily mark required attributes as ``required``
+        for r in required:
+            if r in valid_attributes:
+                valid_attributes[r].required = True
```

Comment:
Just curious on what the long term plan for this is. How would this change?

Reply:
This is the plan! It's just temporary in that they are marked as required at run-time but not saved in the database that way.

Reply:
Gotcha!